### ISSUE TITLE
qemacs won't fully "start up"

### ISSUE THREAD

**Issue by cmsigler** [qemacs/qemacs#11]
Hi,

r1179 works fine for me, but r1182 and r1185 don't.  They hang when running them, appearing not to "start up."  Environment:

Arch Linux, kernel 5.19.13-arch1-1, XFCE desktop (running on X11, not Wayland, obv.), qemacs run in shell terminal.

If I run `qemacs filename.ext' at the command prompt, nothing happens.  I tried to run under X with DISPLAY=:0.0 and also inside the terminal emulator with DISPLAY unset.  No go either way :(  Running inside of gdb gives this short traceback:

Starting program: /usr/bin/qemacs 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
^C
Program received signal SIGINT, Interrupt.
0x000055555557641f in ?? ()
(gdb) where
#0  0x000055555557641f in ?? ()
#1  0x000055555557e23e in qe_register_commands ()
#2  0x000055555558835a in ?? ()
#3  0x00005555555a44db in url_main_loop ()
#4  0x00005555555750bc in main ()
(gdb) quit

I'm going to be away for a couple of days, but please let me know if you need any more info or testing and I'll send ASAP.  HTH :)

Clemmitt Sigler

**Comment by cmsigler** [comment]
Hi,

I apologize for forgetting that the build process produces qe_g and xqe_g with debugging symbols :-\

Thinking I might have something stupid in my ~/.qe/config file I mv'ed that to a different filename.  Then when I run xqe_g or qe_g under gdb I get the following traceback:

====== CUT HERE ======
Using host libthread_db library "/usr/lib/libthread_db.so.1".
^C
Program received signal SIGINT, Interrupt.
0x0000555555576daf in qe_register_binding (nb_keys=10, keys=0x7fffffffd600, d=0x55555563ff40 <basic_commands+32>, m=0x3c) at qe.c:333
333	    while (*lp != NULL && (*lp)->cmd == d)
(gdb) where
#0  0x0000555555576daf in qe_register_binding (nb_keys=10, 
    keys=0x7fffffffd600, d=0x55555563ff40 <basic_commands+32>, m=0x3c)
    at qe.c:333
#1  qe_register_command_bindings (m=m@entry=0x0, 
    d=d@entry=0x55555563ff40 <basic_commands+32>, keystr=<optimized out>)
    at qe.c:370
#2  0x000055555557e74d in qe_register_commands (m=m@entry=0x0, 
    cmds=cmds@entry=0x55555563ff20 <basic_commands>, len=len@entry=140)
    at qe.c:427
#3  0x00005555555887db in qe_init (opaque=opaque@entry=0x7fffffffdfb0)
    at qe.c:9782
#4  0x00005555555a4297 in url_main_loop (
    init=init@entry=0x555555588660 <qe_init>, 
    opaque=opaque@entry=0x7fffffffdfb0) at unix.c:319
#5  0x0000555555575b87 in main (argc=<optimized out>, argv=<optimized out>)
    at qe.c:9941
====== CUT HERE ======

Is the program getting caught in an infinite loop in this while statement in qe_register_binding? ->

    while (*lp != NULL && (*lp)->cmd == d)
        lp = &(*lp)->next;

Am I the only Linux user seeing this problem? 8-0 Again, the last working revision I build is r1179.  HTH, and thank you!

Clemmitt Sigler

**Comment by Piscium** [comment]
i ran with gdb like Clemmitt and got similar results.

^C
Program received signal SIGINT, Interrupt.
qe_register_binding (nb_keys=10, keys=0x7fffffffd700, d=0x55555563dc40 <basic_commands+32>, m=0x55555563dc40 <basic_commands+32>) at qe.c:334
334             lp = &(*lp)->next;
(gdb) where
#0  qe_register_binding (nb_keys=10, keys=0x7fffffffd700, d=0x55555563dc40 <basic_commands+32>, m=0x55555563dc40 <basic_commands+32>) at qe.c:334
#1  qe_register_command_bindings (m=m@entry=0x0, d=d@entry=0x55555563dc40 <basic_commands+32>, keystr=<optimized out>) at qe.c:370
#2  0x000055555557d2de in qe_register_commands (m=m@entry=0x0, cmds=cmds@entry=0x55555563dc20 <basic_commands>, len=len@entry=140) at qe.c:427
#3  0x0000555555587728 in qe_init (opaque=opaque@entry=0x7fffffffe0b0) at qe.c:9782
#4  0x00005555555a3deb in url_main_loop (init=init@entry=0x5555555875a0 <qe_init>, opaque=opaque@entry=0x7fffffffe0b0) at unix.c:319
#5  0x00005555555740bc in main (argc=<optimized out>, argv=<optimized out>) at qe.c:9941

**Comment by cmsigler** [comment]
Hi all,

So, I am a bit puzzled about this problem.  It may be related to XFCE or Arch?  But that seems strange to me.  I decided that since I hadn't updated my installation for a long time maybe I had gotten into an inconsistent state of libraries, etc., so I got everything up-to-date yesterday (Monday, Nov. 7, 2022).  Rebuilding after getting up-to-date didn't help.

Unfortunately, I appear to be incompetent at using git....  I was using 'git reset' but it appears I should have been using 'git checkout' :-\  So I learned how to use 'git bisect' and found the last working version for me is commit c561a42 (r1180).  Starting with commit 0099f13 (r1181) this hang problem starts for me.

HTH in debugging.  Again, this is on Arch under XFCE just as reported by Piscium.  If other Linux users aren't seeing this hang issue I think it would be very valuable to know what distributions, desktops, library revisions, etc., *are* working to see if the failure source can be tracked down.

P.S.:  In the qemacs-git PKGBUILD, use the line

    source=("git+https://github.com/qemacs/qemacs.git#commit=c561a42")

to pull the last working commit and build the corresponding pkg :)

Clemmitt

**Comment by cmsigler** [comment]
Hi all,

@posguy99 -- Very, very interesting that qemacs HEAD builds and runs under Debian stable (Bullseye)....  Thank you for that data point! :-)  Would you please provide more info, such as kernel version, and file and ldd outputs for your qemacs binary?  Thank you for your help! :-)

I am running Arch Linux kernel 6.0.7-arch1-1 (and first saw this issue under 5.19.13-arch1-1).  When I build HEAD, commit 6880277, the outputs of file and ldd for the qemacs binary which installs are:

========

    $ file qemacs
    qemacs: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=ba59d24e6d149c3374f3de9bf6a52fac138dd8ad, for GNU/Linux 4.4.0, stripped

    $ ldd qemacs
    	linux-vdso.so.1 (0x00007ffcc7ea9000)
    	libm.so.6 => /usr/lib/libm.so.6 (0x00007fbbad302000)
    	libXv.so.1 => /usr/lib/libXv.so.1 (0x00007fbbad2fa000)
    	libXext.so.6 => /usr/lib/libXext.so.6 (0x00007fbbad2e5000)
    	libX11.so.6 => /usr/lib/libX11.so.6 (0x00007fbbad1a2000)
    	libc.so.6 => /usr/lib/libc.so.6 (0x00007fbbacfbb000)
    	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fbbad525000)
    	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007fbbacf8e000)
    	libXau.so.6 => /usr/lib/libXau.so.6 (0x00007fbbacf89000)
    	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007fbbacf81000)

========

I'm obviously I'm grasping at straws here.  Would appreciate any help anyone following can provide.  TIA!

Clemmitt

**Comment by posguy99** [comment]
Here you go.

$ file /usr/local/bin/qemacs
/usr/local/bin/qemacs: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-li
nux-x86-64.so.2, BuildID[sha1]=d4424143198619e59c7231cd33e1ae74249a75c8, for GNU/Linux 3.2.0, stripped

$ ldd /usr/local/bin/qemacs
linux-vdso.so.1 (0x00007ffe8eccb000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc721538000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc7213f4000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc72121f000)
/lib64/ld-linux-x86-64.so.2 (0x00007fc721546000)

**Comment by cmsigler** [comment]
@posguy99 ,

Thank you.  Obvious first note is that your qemacs is terminal only, built without X support.  But this doesn't seem to help me.... When I try to run qe_g under gdb, which links with the same libs as yours (except my newer Linux version doesn't explicitly link against libdl.so.2) and has no X, it gets caught in an infinite loop just like xqe_g.

Also, not sure about this but your linking is "for GNU/Linux 3.2.0," while mine is later, 4.4.0.  I'm not knowledgeable about what difference that might make.

Would you please let me know what kernel version you're running?  The output of 'uname -a' would be helpful :-)

One other thing, if it's not too much trouble.  Would you please try compressing then uploading as an attachment your /usr/local/bin/qemacs binary -- not sure what to use these days, zstd, xz, bzip2, gzip, just zip???  I would be interested in seeing if it runs on my systems.  Not sure if this is valuable info, so no worries if it's too much trouble.  Mainly for my curiosity....

It's been a very, very long time since I did any actual debugging.  I think I will have to dig in and give it a try.  I probably ought to set a breakpoint where the program calls the function qe_register_binding() in qe.c, then step forward from there....

Thank you again :-)

Clemmitt

**Comment by posguy99** [comment]
Certainly.

```
$ uname -a 
Linux debian-linode.local 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
```

**Comment by cmsigler** [comment]
@posguy99 ,

Would you mind testing the following command line operations, please?  If not, no worries.  I am hoping to make sure we are testing the exact same thing:

    $ mkdir ~/some_tmp_dir
    $ cd ~/some_tmp_dir
    $ git clone https://github.com/qemacs/qemacs.git
    $ cd qemacs
    $ git reflog
    $ ./configure
    $ make
    $ gdb ./qe_g
    (gdb) r

If qemacs starts normally, that's what I'd like to confirm.  Please exit qemacs and you're finished.  If it doesn't start up, please do:

    ^C      [ hit Ctrl-C ]
    (gdb) where
    (gdb) q
    Quit anyway? (y or n) y
    $

As you are able, please post the output from both 'git reflog' and 'where'.  And thank you once again for your help! :-D

Clemmitt

**Comment by posguy99** [comment]
Qemacs starts up normally.

```
$ git reflog 
6880277 (HEAD -> master, origin/master, origin/HEAD) HEAD@{0}: clone: from https://github.com/qemacs/qemacs.git

```

**Comment by cmsigler** [comment]
@posguy99 ,

Thank you!

Well, I'm stumped.  For some seemingly obscure reason, builds on Arch get caught in an infinite loop in lines 333 & 334 of qe.c:

    while (*lp != NULL && (*lp)->cmd == d)
        lp = &(*lp)->next;

In running gdb on my binary I notice that some variables report as "optimized out" which makes me curious as to whether gcc is doing some bad stuff in optimization.  My gcc version is:

    $ gcc --version
    gcc (GCC) 12.2.0

Would you please post your gcc version?  Thank you again.  Your info is indeed helpful :-)  I'm just not knowledgeable enough to recognize the (obvious?) cause.... :-\

Clemmitt

**Comment by posguy99** [comment]
Don't have gcc installed.

```
$ cc --version 
Debian clang version 11.0.1-2
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by cmsigler** [comment]
@posguy99 ,

Interesting....  You're building with clang.  OK, I need to install and test building with clang to see what that does.  That's a big help.  Thank you :-D

Clemmitt

**Comment by posguy99** [comment]
My config file is here, I doubt this has anything do do with anything, but for completeness...

https://github.com/posguy99/dotfiles/blob/master/.qe/config

**Comment by cmsigler** [comment]
Hi,

@posguy99 , I had nuked my config file early in the debugging process just to make sure I was starting from ground zero.

=>
=> AMAZING!!!  This is almost certainly caused by a gcc bug (which must be obscure?).  All I needed to do was install clang, then in the freshly cloned qemacs top-level dir I ran './configure --cc=clang' followed by 'make' and the programs built then ran perfectly normally.
=>

So -- ***NOTE*** -- Apparent bug in "gcc (GCC) 12.2.0", pkg "gcc 12.2.0-1 (base-devel)", on Arch Linux causes commits after c561a42 to fail to run (because the program gets caught in an infinite loop).

Closing this bug.  My thanks to everyone, and especially to posguy99, for your help! :-D

Clemmitt

**Comment by Piscium** [comment]
Hmmm. Over the years a few times I found strange things happening and thought I had found a compiler. On further inspection it was never the case. Of course there is such a thing as compiler bugs, but the chances of finding one in normal development are not that big. There is usually another reason for the issue. In this case I have no idea where the problem lies, but I would guess that there is in qemacs code something that is considered by the compiler as undefined behaviour so it works with some compilers but not others, or with some libraries and not others. So in my opinion it is premature to close this issue.

Although I don't want to spend much time on this, I did find some time to start my Debian bullseye VM, installed gcc which in bullseye is version 10, and it does not hang following the instructions above. I also have a Debian sid VM, if I find the time I will do the same exercise on it.

**Comment by Piscium** [comment]
Another test this time on Debian Sid. It hangs. Sid has gcc 12.2 which is the same version as Arch. So it may be that gcc 12.2 optimises away some code that causes the hang.

**Comment by chqrlie** [comment]
Hello,

This is my observation too, under macOS, Debian, freeBSD and openBSD, but my systems are probably outdated.

The offending code runs an infinite loop trying to skip entries in a linked list.  Therefore, for some reason, the list on the target system becomes corrupted and has a back link.

I suspect a compiler issue, maybe related to over-optimisation of code assumed to have undefined behavior, such as copying to the flexible array at the end of the structure, which is defined as an array of 1 byte for compatibility with pre-C99 compilers.

I shall set up a system with the Debian version that shows the problem and investigate.

Best regards.

Chqrlie.

PS: sorry for not reacting sooner, I was travelling on a busy schedule :)

**Comment by qemacs** [comment]
Dear all,

I fully agree.

I am not suggesting a compiler bug, but possibly an unexpected compiler behavior on code that may have undefined behavior as per the C Standard.
Such problems do not go away without a proper fix in the source code, but investigating them requires access to a system where they cause problems.
I shall do this over the week-end.

Best regards.

Chqrlie.

**Comment by cmsigler** [comment]
Dear all,

I did my own stupid compiler tricks test.  I took the git clone, made a copy with 'cp -ia', then built one using the default compiler, gcc, and the other using clang.  gcc 12.2.0 and clang 14.0.6.  I ran 'gdb xqe_g' with both.  I set a breakpoint at line 333 of qe.c via 'b qe.c:333'.

Already interesting things happened.  The gcc run set the breakpoint without complaint.  The clang run gave this:

    (gdb) b qe.c:333
    No source file named qe.c.
    Make breakpoint pending on future shared library load? (y or [n]) y
    Breakpoint 1 (qe.c:333) pending.

Running the clang debug **never** **reaches** **line** **333** at all, according to gdb.  xqe_g starts normally and pops open an X window on my desktop.  So, is everything just aggressively optimized away by clang when it is used with no special options to compile qemacs???  I know diddly about compilers and debuggers at a low level :-\

Running the gcc debug stops at line 333 the first time, in which case *lp == NULL is true.  Continuing, it stops at line 333 with:

    (gdb) p *lp
    $3 = (KeyDef *) 0x555555658fb0
    (gdb) p (*lp)->cmd
    $4 = (const struct CmdDef *) 0x55555563ff40 <basic_commands>
    (gdb) p d
    $5 = (const CmdDef *) 0x55555563ff60 <basic_commands+32>

Continuing, it stops at line 333 with:

    (gdb) p *lp
    $6 = (KeyDef *) 0x555555658fd0
    (gdb) p (*lp)->cmd
    $7 = (const struct CmdDef *) 0x55555563ff60 <basic_commands+32>
    (gdb) p d
    $8 = (const CmdDef *) 0x55555563ff60 <basic_commands+32>

And this is the beginning of the infinite looping.

Doing some more snooping on the gcc build, I started from scratch and set a breakpoint at line 330, but gdb acted "weird."  It didn't stop at 330 but instead at 333, perhaps due to some optimization?  (I'm no expert.)  I then stepped forward.  Control kicks out to qe_register_command_bindings, then goes into qe_register_commands, then qe_register_command_bindings.  It then seems to go to qe_register_binding but again I think it's acting "weird?"  My results:

    qe_register_command_bindings (m=m@entry=0x0, d=d@entry=0x55555563ff60 <basic_commands+32>, keystr=0x5555555fb936 "M-#, C-x 8 RET") at qe.c:363
    363	{
    (gdb) s
    366	    const char *p = keystr;
    (gdb) s
    368	    while (p && *p) {
    (gdb) s
    370	        res = qe_register_binding(m, d, keys, nb_keys);
    (gdb) s
    0x0000555555576c22 in qe_register_binding (nb_keys=<optimized out>, 
        keys=<optimized out>, d=<optimized out>, m=<optimized out>) at qe.c:330
    330	    lp = m ? &m->first_key : &qs->first_key;

Maybe I'm totally not understanding, but why does it enter into qe_register_binding at line 330?  Shouldn't it enter at line 315?  Is this some effect of compiler optimization, or possibly an optimization bug???  And do the "optimized out" variable values indicate trouble brewing? 8-\

@Piscium , I totally agree about actual compiler bugs being vanishingly infrequent.  And I really don't know diddly.  I would not expect that I would be the person to stumble over such a thing....

Chqrlie, thank you for working on this and for producing qemacs for all of us end-user schlubs ;-)

Clemmitt

**Comment by Piscium** [comment]
qemacs (latest in git) also hangs on Fedora 36 with gcc 12.2. In a nutshell, based on all the tests the conclusion is that qemacs hangs with gcc 12.2, but not with gcc 10 nor clang. So it seems that kernel, libraries or desktop environments are not relevant with respect to reproducing the issue, what matters is the compiler version.

**Comment by chqrlie** [comment]
Dear all,

Yes indeed the bug is related to the gcc 12.2 optimizer (and possibly earlier versions)

yes it is a NASTY BUG

As I was suspecting, my code has a very subtle kind of undefined behavior that triggers the optimizer to generate code that causes an infinite loop.

The offending code is in util.c, in functions strtokey1 and utf8_decode.

It first appeared in git commit 0099f138c9af51d26b9e331780db17ccb00f3cab when I moved and modified the code for utf8_decode from charset.c to util.c

strtokey1 has this calling sequence:

    } else {
        key = utf8_decode(&p);
        *pp = p;
    }
    return key;

The code calls utf8_decode, storing `p` (register %rbp) to (%rsp) and %r12, passing the value of %rsp as the argument `&p`, which is fine, but after the call, it jumps to exit code that uses the previous value of `p` stored in `%r12` to store it to `*pp` (`pp` is stored in %r13):

The assembly code can be produced by `make util.s`

It seems gcc 12.2 assumes utf8_decode does not modify p. The prototype is `char32_t utf8_decode(const char **pp)` which can modify `*pp`, so it should not make this assumption, but since the source code for function `utf8_decode` is in the same file now, gcc can inspect it for purposes of inlining and other optimisation attempts.

utf8_decode uses a const unsigned char pointer `p` to parse the string as 8-bit bytes with unsigned semantics (no sign extension). I use such a type for the pointer to avoid extra casts in case `char` happens to be a signed type (on targets without -funsigned-char` semantics).

I define and initlialize `p` with:

    const u8 *p = *(const u8 **)pp;

Then parse the bytes, converting UTF-8 encoded values with lax extended semantics.

Finally, I store the updated value of `p` to `*pp` with:

*(const u8 **)pp = p;

This is a common idiom for C programmers: I have a pointer, cast it to a different type and store a value via this alternate pointer type.

But technically, THIS HAS UNDEFINED BEHAVIOR BECAUSE IT VIOLATES THE STRICT ALIASING RULE!!!!

And gcc 12.2 generates the expected code, but does not flag `*pp` as having been modified by this line.

As a consequence it flags `utf8_decode` as a pure function meaning that `utf8_decode(&p)` does not modify `p`…
hence it may generate code that uses the previous value of `p` in `strtokey1()`.

I have been programming in C for 40 years, have written millions of lines and consider myself fairly advanced in my understanding of the C language, and like many fellows before me, I got got bitten by this nasty compiler optimisation.  C semantics are already too subtle for most programmers to cope with… for compiler writers to assume every program is 100% compliant or else make it crash is INSANE.  

I shall commit a fix using different casts which have defined semantics:

    const u8 *p = (const u8 *)*pp;

and 

   *pp = (const char *)p;

And I shall of course fix similar problems in the rest of the code.

Best regards

Chqrlie.
